Modules/constants.c

```diff
@@ -48,38 +49,81 @@ LDAPerr(int errnum)
 
 /* Convert an LDAP error into an informative python exception */
 PyObject *
-LDAPerror(LDAP *l, char *msg)
+LDAPraise_for_message(LDAP *l, char *msg, LDAPMessage *m)
 {
     if (l == NULL) {
         PyErr_SetFromErrno(LDAPexception_class);
+        ldap_msgfree(m);
         return NULL;
     }
     else {
-        int myerrno, errnum, opt_errnum;
+        int myerrno, errnum, opt_errnum, msgid = -1, msgtype = 0;
         PyObject *errobj;
         PyObject *info;
         PyObject *str;
         PyObject *pyerrno;
-        char *matched, *error;
+        PyObject *pyresult;
+        PyObject *pyctrls = NULL;
+        char *matched = NULL,
+             *error = NULL,
+             **refs = NULL;
+        LDAPControl **serverctrls = NULL;
 
         /* at first save errno for later use before it gets overwritten by another call */
         myerrno = errno;
 
-        opt_errnum = ldap_get_option(l, LDAP_OPT_ERROR_NUMBER, &errnum);
-        if (opt_errnum != LDAP_OPT_SUCCESS)
-            errnum = opt_errnum;
+        if (m != NULL) {
+            msgid = ldap_msgid(m);
+            msgtype = ldap_msgtype(m);
+            ldap_parse_result(l, m, &errnum, &matched, &error, &refs,
+                              &serverctrls, 1);
+        }
 
-        if (errnum == LDAP_NO_MEMORY)
-            return PyErr_NoMemory();
+        if (msgtype <= 0) {
+            opt_errnum = ldap_get_option(l, LDAP_OPT_ERROR_NUMBER, &errnum);
+            if (opt_errnum != LDAP_OPT_SUCCESS)
+                errnum = opt_errnum;
+
+            if (errnum == LDAP_NO_MEMORY) {
+                return PyErr_NoMemory();
+            }
+
+            ldap_get_option(l, LDAP_OPT_MATCHED_DN, &matched);
+            ldap_get_option(l, LDAP_OPT_ERROR_STRING, &error);
+        }
 
         if (errnum >= LDAP_ERROR_MIN && errnum <= LDAP_ERROR_MAX)
             errobj = errobjects[errnum + LDAP_ERROR_OFFSET];
         else
             errobj = LDAPexception_class;
 
         info = PyDict_New();
-        if (info == NULL)
+        if (info == NULL) {
+            ldap_memfree(matched);
+            ldap_memfree(error);
+            ldap_memvfree((void **)refs);
+            ldap_controls_free(serverctrls);
             return NULL;
+        }
+
+        if (msgtype > 0) {
+            pyresult = PyInt_FromLong(msgtype);
+            if (pyresult)
+                PyDict_SetItemString(info, "msgtype", pyresult);
+            Py_XDECREF(pyresult);
+        }
+
+        if (msgid >= 0) {
+            pyresult = PyInt_FromLong(msgid);
+            if (pyresult)
+                PyDict_SetItemString(info, "msgid", pyresult);
+            Py_XDECREF(pyresult);
+        }
+
+        pyresult = PyInt_FromLong(errnum);
+        if (pyresult)
+            PyDict_SetItemString(info, "result", pyresult);
+        Py_XDECREF(pyresult);
```

Comment:
This is missing error handling; `PyInt_FromLong` and `PyDict_SetItemString` can fail.
Same for other Python C-API below.

Reply:
Is there an example how this should be done? All I can find is both of them treated the this way across the codebase.

Reply:
You are right. I'll send a separate PR to add error handling, so this isn't blocked.

Reply:
For reference, it's here: https://github.com/python-ldap/python-ldap/pull/341